devito/profiler.py

```diff
@@ -39,50 +56,198 @@ def add_profiling(self, block, name, omp_flag=None):
                            "(double)(end_%(n)s.tv_sec-start_%(n)s.tv_sec)+" +
                            "(double)(end_%(n)s.tv_usec-start_%(n)s.tv_usec)" +
                            "/1000000") %
-                          {"sn": self.name, "n": name}),
+                          {"sn": self.t_name, "n": name}),
             ])
         ]
 
-        return init + block + end
+        return init + code + end
 
-    @property
-    def get_timings_class(self):
-        """Returns a :class:`ctypes.Structure` subclass defining our structure
+    def get_oi_and_flops(self, name, code, size):
+        """Calculates the total operation intensity of the code provided.
+        If needed, lets the C code calculate it.
 
-        :returns: A :class:`Timings` class definition
+        :param name: The name of the field to be populated
+        :param code: The code to be profiled.
         """
-        return type("Timings", (Structure,), {"_fields_": self.fields})
+        if name not in self.oi:
+            self.oi[name] = 0
+        if name not in self.flops_defaults:
+            self.flops_defaults[name] = 0
+
+        loads = {"stores": 0}
+
+        for elem in code:
+            if isinstance(elem, Assign):
+                assign_flops = self._get_assign_flops(elem, loads)
+                self.oi[name] += assign_flops
+                self.flops_defaults[name] += assign_flops
+            elif isinstance(elem, For):
+                for_flops = self._get_for_flops(name, elem, loads)
+                self.oi[name] += for_flops
+            elif isinstance(elem, Block):
+                block_oi, block_flops = self._get_block_oi_and_flops(name, elem, loads)
+                self.oi[name] += block_oi
+                self.flops_defaults[name] += block_flops
+
+        self.oi[name] = float(self.oi[name]) / (size*(len(loads) + loads["stores"] - 1))
+
+    def _get_for_flops(self, name, loop, loads):
+        loop_flops = 0
+        loop_oi_f = 0
+
+        if isinstance(loop.body, Assign):
+            loop_flops = self._get_assign_flops(loop.body, loads)
+            loop_oi_f = loop_flops
+        elif isinstance(loop.body, Block):
+            loop_oi_f, loop_flops = self._get_block_oi_and_flops(name, loop.body, loads)
+        elif isinstance(loop.body, For):
+            loop_oi_f = self._get_for_flops(name, loop.body, loads)
+
+        flops_calc_stmt = Statement("%s->%s += %f" % (self.f_name, name, loop_flops))
```

Comment:
We could short-cut here to avoid `flops += 0.0` statements. Just a suggestion to make the generated code easier to read.